File: 백준/Silver/15663. N과 M （9）/N과 M （9）.js
/*
이전 값이 아니라 같은 형제들 중에 이미 선택되었다면 걸린다. 

그러면 형제들 중에 이미 같은 값이 선택되었다는 사실은 어떻게 확인할 수 있지?

*/

const filePath = process.platform === 'linux' ? '/dev/stdin' : 'e.txt';
const input = require('fs').readFileSync(filePath).toString().trim().split('\n');

const [count, maxLength] = input[0].split(' ').map(Number);

const numbers = input[1]
  .split(' ')
  .map(Number)
  .sort((a, b) => a - b);

const temp = [];
const resultList = [];
const visited = new Array(count).fill(false);

function bt() {
  if (temp.length === maxLength) {
    resultList.push(temp.join(' '));
    return;
  }

  let value = null;
  for (let i = 0; i < count; i++) {
    if (visited[i] === true) continue; // 이미 temp에 추가된 숫자
    if (value === numbers[i]) continue;

    value = numbers[i];
    visited[i] = true;
    temp.push(numbers[i]);
    bt();
    visited[i] = false;
    temp.pop();
  }
}

bt();

console.log(resultList.join('\n'));

File: 백준/Silver/15663. N과 M （9）/N과 M （9）.py
'''
그 차례에 이미 그 수가 등록되었다면 continue
'''

n, m = map(int, input().split())
lst = list(map(int, input().split()))

lst.sort()

temp = []
prevNum = None
def perm (curIdx, selected):
    if len(temp) == m:
        print(' '.join(map(str, temp)))
        return;
    
    prev = None
    for i in range(n):
        if i in selected: # 중복 선택이 안되므로 이미 선택된 값인지 확인ㅇ
            continue
                
        cur = lst[i]
        
        if prev == cur: # 정렬된 상태이기 떄문에 이전 값과 동일하다면 pass
            continue;

        prev = cur
        selected.append(i)
        temp.append(cur)
        
        perm(i + 1, selected)
        temp.pop()
        selected.pop()
            
        
perm(0, [])
            
    